client/state/data-layer/wpcom/read/tags/utils.js

```diff
@@ -0,0 +1,40 @@
+/**
+ * External dependencies
+ */
+import { map, compact, concat, isObject, isArray } from 'lodash';
+import { decodeEntities } from 'lib/formatting';
+
+/**
+ * Normalize response from the api so whether we get back a single tag or a list of tags
+ * we always pass forward a list
+ * Also transform the api response to be something more calypso-friendly
+ *
+ * @param  {Tag|Tags} apiResponse api response from the tags endpoint
+ * @return {Tags} An object containing all of the normalized tags in the format:
+ *  [
+ *    { id, displayName, url, title, slug },
+ *    ...
+ *  ]
+ */
+export function fromApi( apiResponse ) {
+	if ( ! apiResponse || ( ! apiResponse.tag && ! apiResponse.tags ) ) {
+		if ( process.env.NODE_ENV === 'development' ) {
+			throw new Error( 'bad api response for /read/tags' );
+		}
+		return [];
+	}
+
+	const tags = compact( concat(
+		[],
+		isObject( apiResponse.tag ) && apiResponse.tag,
+		isArray( apiResponse.tags ) && apiResponse.tags,
+	) );
+
+	return map( tags, tag => ( {
+		id: tag.ID,
+		displayName: decodeEntities( tag.display_name ),
+		url: `/tag/${ tag.slug }`,
+		title: decodeEntities( tag.title ),
+		slug: tag.slug.toLowerCase(),
+	} ) );
+}
```

Comment:
> meh about altering it to use camelCase over the normal API shape. It's CPU time, memory churn, and bytes on the wire we don't really need to ship.

@blowery thanks for raising this concern. I use the snake_case names as a canary which indicates that we haven't made an attempt to guard data coming into Calypso. that data flows through the app and ends up feeling very strange to me: it's not accounted for anywhere until the leaf component which just assumes its there; when things on the API change Calypso stops working and there's no indication why; the naming and data shapes look entirely different than data generated from within Calypso.

further I'm more concerned about leading to actual bugs than risking performance but in the data layer I would argue that performance is of no practical value. how long does a network request take? how long does it take to rearrange and rename data? if we get twenty network requests each minute and normalizing our data to ensure the safety of the app takes an additional 5ms (being generous) each then those six frames we lose each minute certainly wouldn't be noticeable to the customer would they?

> weird use of tap

i'm exploring here. this is something I like about the data-layer - it's all async which let's us `throw` the way exceptions are design to work: if something isn't right, stop trying to continue using it. we'll see in the console that something is wrong at least. I can accept your concern about not handling that failure in such a way that informs the customer though.

also, what's weird about it?